Keda cron scaler (#509)
SAP-samples/kyma-runtime-extension-samples: keda-cron-scaler/README.md

![off-work](assets/keda-scale-off-work.png)

## Scenario

Lets put cron based scaler to action.

Assume we have a **development cluster** where we want to run workloads only during work hours.

Lets say **Monday - Friday, 8 AM to 6 PM**

### Prerequisites

- [SAP BTP, Kyma runtime instance](../prerequisites/#kyma)
- [Kubernetes tooling](../prerequisites/#kubernetes)
- [KEDA and Serverless Modules enabled in Kyma]((https://help.sap.com/docs/btp/sap-business-technology-platform/enable-and-disable-kyma-module))

### Steps

- Export environment variables

```shell
export NS={your-namespace}
```

- Create namespace and enable istio injection if not already done

```shell
kubectl create ns ${NS}
# only required once to enable istio sidecar. Ignore if done already
kubectl label namespaces ${NS} istio-injection=enabled
```

- Create sample workloads. One deployment and one function

```shell
kubectl -n ${NS} apply -f k8s/deployment.yaml
kubectl -n ${NS} apply -f k8s/function.yaml
```

- Apply the KEDA cron based scaling to these workloads

```shell
kubectl -n ${NS} apply -f k8s/keda-cron-scaler.yaml
```

### How it works

> KEDA `scaledobject` resource can be configured with a trigger of type cron.
> 
> In the cron scaler, we can then specify to have the workloads running only during the working hours.

Reviewer: ```suggestion
The KEDA `scaledobject` resource can be configured with a trigger of type `cron`. Within the cron scaler, you can specify that the workloads should only run during working hours.
```

```yaml
  triggers:
    - type: cron
      metadata:
        # The acceptable values would be a value from the IANA Time Zone Database.
        timezone: Europe/Berlin  
        # At 08:00 AM, Monday through Friday
        start: 0 8 * * 1-5
        # At 06:00 PM, Monday through Friday
        end: 0 18 * * 1-5
        # ie. Your MINIMUM replica count for this workload
        desiredReplicas: "1"
```

For each type of workload, the **scaleTargetRef** need to be specified

```yaml
spec:
  scaleTargetRef:
    apiVersion: apps/v1
    kind: Deployment
    name: test-keda-cron-nginx
```

```yaml
spec:
  scaleTargetRef:
    apiVersion: serverless.kyma-project.io/v1alpha2
    kind: Function
    name: test-keda-cron-function
```

### View the events

> Check the events during the trigger start or end time. Here you can see KEDA scaling down the replicas

Reviewer: ```suggestion
Check the events occurring at the start or end time of the trigger. In the following example, you can see KEDA scaling down the replicas:
```

```shell
kubectl -n ${NS} get events
```

```shell
LAST SEEN   TYPE      REASON                       OBJECT                                       MESSAGE
7m34s       Normal    Killing                      pod/test-keda-cron-nginx-86b78b79df-r42zd    Stopping container istio-proxy
7m34s       Normal    Killing                      pod/test-keda-cron-nginx-86b78b79df-r42zd    Stopping container nginx
7m31s       Warning   Unhealthy                    pod/test-keda-cron-nginx-86b78b79df-r42zd    Readiness probe failed: HTTP probe failed with statuscode: 503
7m34s       Normal    SuccessfulDelete             replicaset/test-keda-cron-nginx-86b78b79df   Deleted pod: test-keda-cron-nginx-86b78b79df-r42zd
7m34s       Normal    KEDAScaleTargetDeactivated   scaledobject/test-keda-cron-nginx            Deactivated apps/v1.Deployment demos/test-keda-cron-nginx from 1 to 0
7m34s       Normal    ScalingReplicaSet            deployment/test-keda-cron-nginx              Scaled down replica set test-keda-cron-nginx-86b78b79df to 0 from 1
```

> > Note: Events are only available until after 1 hour of trigger.

Reviewer: ```suggestion
> Note: Events are only available for one hour after the trigger.
```

> ### First hand experience

Reviewer: ```suggestion
### First-hand experience
```

> I applied the KEDA cron scaler to all my custom workloads in my Kyma cluster.

Reviewer: ```suggestion
I applied the KEDA cron scaler to all custom workloads in my Kyma cluster.
```

> All my deployment replicas scaled down to zero

Reviewer: ```suggestion
All my deployment replicas were scaled down to zero.
```

![dep-off](assets/keda-off-hours.png)

> and number of nodes (VMs) reduced from 4 to 3

Reviewer: ```suggestion
Additionally, the number of nodes (VMs) was reduced from 4 to 3.
```
